### PR TITLE
Shared Regs functional with test programs

### PR DESCRIPTION
**Pull request by ThomasSeaver**
Altered VirtualMachine.cs to extend reg num, add propagation call to data manipulation instructions, and add method to update shared regs from within vm

Altered BotManager.cs to add coroutine that, after a delay, uses the method in bot.VM to update shared regs across teams

Added SharedRegisters and SharedRegisters2 as testing programs that demonstrate the bots waiting for eachother to update the regs. You can examine the programs yourself, but they basically are taking turns changing the reg and then waiting for their partner to change the reg, moving each time they set a reg for visual clarity.

I did not edit the editor's display of regs since that is probably something we should discuss beforehand and the system extends to 10 regs automatically via the numreg change. 

I exposed the propagation delay variable in the inspector, but the coroutine seems to use the first value in the variable even after changing it in the inspector? I'm really not sure what's going on there; even after changing it in the inspector, the debug log would report the original value. I'm not even sure it was using the inspector value at all, and I think I would reset to 0 unless I set the delay manually in the awake function. Regardless, the delay functions, the coroutine just doesn't seem to want to use any updated version of the variable, and I'm not sure what part of C# I'm running up against there.

### CODE REVIEW

File: Assets/Decontainment/Singletons/BotManager.cs [hunk 1]
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
         Vector2 look = target.transform.position - targeter.transform.position;
         return (int)Vector2.SignedAngle(targeter.transform.right, look);
     }
+
+    public IEnumerator PropagationCoroutine (int registerNumber, int registerValue, int botTeamID) {
```
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]:
```suggestion
    public IEnumerator PropagationCoroutine(int registerNumber, int registerValue, int botTeamID)
    {
```
Stay consistent with the rest of the codebase

File: Assets/Decontainment/Assembly/VirtualMachine.cs [hunk 2]
```diff
@@ -164,6 +171,10 @@ public void Tick()
         OnTick?.Invoke();
     }
 
+    public void updateSharedReg(int registerNumber, int registerValue) {
```
Comment (BrandonShockley) [inline Assets/Decontainment/Assembly/VirtualMachine.cs]:
```suggestion
    public void UpdateSharedReg(int registerNumber, int registerValue)
    {
```
Keep the code style consistent

File: Assets/Decontainment/Singletons/BotManager.cs [hunk 3]
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
...
+        //Debug.Log("Propagating with delay: " + propagationDelay);
```
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]: Don't leave commented out code in the PR

File: Assets/Decontainment/Singletons/BotManager.cs [hunk 4]
```diff
@@ -71,4 +75,14 @@ public int GetTargetHeading(Controller targeter, int targetIndex)
...
+        yield return new WaitForSecondsRealtime(propagationDelay);
```
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]: Change this to use WaitForSeconds

File: Assets/Decontainment/Assembly/VirtualMachine.cs (hunk 2 above)
Comment (BrandonShockley) [inline Assets/Decontainment/Assembly/VirtualMachine.cs]: Also this method isn't used anywhere

File: Assets/Decontainment/Singletons/BotManager.cs [hunk 5]
```diff
@@ -15,6 +16,8 @@ public enum DistanceType
         FARTHEST,
     }
 
+    public float propagationDelay;
```
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]: Make this a private variable with the `[SerializeField]` attribute

File: Assets/Decontainment/Singletons/BotManager.cs (hunk 1 above)
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]: I highly-highly recommend making the coroutine method private and making a `public void PropagateRegister` method that sets up the coroutine for you. That way you don't have to type as much code to do the propagation like you did in the VirtualMachine class.

File: Assets/Decontainment/Assembly/VirtualMachine.cs [hunk 6]
```diff
@@ -6,7 +6,7 @@
 
 public class VirtualMachine
 {
-    public const int NUM_REGS = 5;
+    public const int NUM_REGS = 10;
```
Comment (BrandonShockley) [inline Assets/Decontainment/Assembly/VirtualMachine.cs]: Do what we discussed in the chat

**Review by BrandonShockley — CHANGES_REQUESTED** [review]

**Comment by ThomasSeaver** [comment]
Updated as requested in regards to style, and added new register group to the editor with new color as to visually differentiate. Still functional as before.

File: Assets/Decontainment/Singletons/BotManager.cs [hunk 7]
```diff
@@ -26,6 +30,7 @@ public enum DistanceType
 
     void Awake()
     {
+        propagationDelay = 3.0f;
```
Comment (BrandonShockley) [inline Assets/Decontainment/Singletons/BotManager.cs]: Remove this

**Review by BrandonShockley — CHANGES_REQUESTED** [review]

**Review by BrandonShockley — APPROVED** [review]

**Pull request merged**